marlowe-playground-client/src/MarloweEditor/State.purs

```diff
@@ -0,0 +1,222 @@
+module MarloweEditor.State
+  ( handleAction
+  , editorGetValue
+  , {- FIXME: this should be an action -} editorResize
+  ) where
+
+import Prelude hiding (div)
+import Control.Monad.Except (ExceptT, lift, runExceptT)
+import Control.Monad.Maybe.Extra (hoistMaybe)
+import Control.Monad.Maybe.Trans (MaybeT(..), runMaybeT)
+import Control.Monad.Reader (runReaderT)
+import Data.Array (filter)
+import Data.Either (Either(..), hush)
+import Data.Foldable (for_)
+import Data.Lens (assign, preview, set, use)
+import Data.Lens.Index (ix)
+import Data.Maybe (Maybe(..))
+import Data.String (codePointFromChar)
+import Data.String as String
+import Data.Traversable (traverse)
+import Data.Tuple (Tuple(..))
+import Effect.Aff.Class (class MonadAff, liftAff)
+import Effect.Class (class MonadEffect)
+import FileEvents (readFileFromDragEvent)
+import FileEvents as FileEvents
+import Halogen (HalogenM, liftEffect, modify_, query)
+import Halogen.Monaco (Message(..), Query(..)) as Monaco
+import LocalStorage as LocalStorage
+import MainFrame.Types (ChildSlots, _hasUnsavedChanges', _marloweEditorPageSlot)
+import Marlowe (SPParams_)
+import Marlowe as Server
+import Marlowe.Holes (fromTerm)
+import Marlowe.Linter as Linter
+import Marlowe.Monaco (updateAdditionalContext)
+import Marlowe.Parser (parseContract)
+import Marlowe.Semantics (Contract, emptyState)
+import Marlowe.Symbolic.Types.Request as MSReq
+import MarloweEditor.Types (Action(..), AnalysisState(..), State, _analysisState, _bottomPanelView, _editorErrors, _editorWarnings, _keybindings, _selectedHole, _showBottomPanel)
+import Monaco (IMarker, isError, isWarning)
+import Network.RemoteData (RemoteData(..))
+import Network.RemoteData as RemoteData
+import StaticAnalysis.Reachability (getUnreachableContracts, startReachabilityAnalysis)
+import Servant.PureScript.Ajax (AjaxError)
+import Servant.PureScript.Settings (SPSettings_)
+import StaticData (marloweBufferLocalStorageKey)
+import StaticData as StaticData
+import Text.Pretty (pretty)
+import Types (WebData)
+
+handleAction ::
+  forall m.
+  MonadAff m =>
+  SPSettings_ SPParams_ ->
+  Action ->
+  HalogenM State Action ChildSlots Void m Unit
+handleAction _ (ChangeKeyBindings bindings) = do
+  assign _keybindings bindings
+  void $ query _marloweEditorPageSlot unit (Monaco.SetKeyBindings bindings unit)
+
+handleAction _ (HandleEditorMessage (Monaco.TextChanged text)) = do
+  modify_
+    ( set _selectedHole Nothing
+        <<< set _hasUnsavedChanges' true
+    )
+  liftEffect $ LocalStorage.setItem marloweBufferLocalStorageKey text
+  analysisState <- use _analysisState
+  let
+    parsedContract = parseContract text
+
+    unreachableContracts = getUnreachableContracts analysisState
+
+    (Tuple markerData additionalContext) = Linter.markers unreachableContracts parsedContract
+  markers <- query _marloweEditorPageSlot unit (Monaco.SetModelMarkers markerData identity)
+  void $ traverse editorSetMarkers markers
```

Comment:
`traverse_` instead of `void <<< traverse`

Reply:
done